src/core/page_usage_stats_test.cc

```diff
@@ -426,3 +564,138 @@ TEST_F(PageUsageStatsTest, DefragmentIndexWithNonDefragmentableFields) {
   EXPECT_FALSE(result.quota_depleted);
   EXPECT_EQ(result.objects_moved, 0);
 }
+
+TEST_F(PageUsageStatsTest, DefragReducesWaste) {
+  // This test works with actual defragmentation, by deleting every other json object which creates
+  // holes in pages which cannot be directly freed. The test asserts that wasted memory goes down as
+  // well as committed memory after defragmentation.
+
+  std::vector<std::optional<JsonType>> all_objects;
+
+  constexpr auto total_json = 100;
+  all_objects.reserve(total_json);
+
+  for (auto i = 0; i < total_json; ++i) {
+    auto parsed = ParseJsonUsingShardHeap(GenerateTestJSON(500));
+    EXPECT_TRUE(parsed.has_value());
+    all_objects.emplace_back(std::move(parsed.value()));
+  }
+
+  // Delete every other object to create gaps, so that the pages are partially used.
+  for (size_t i = 0; i < all_objects.size(); i += 2) {
+    all_objects[i].reset();
+  }
+
+  // Allow mimalloc to free any completely empty pages, if any
+  mi_heap_collect(m_.heap(), true);
+
+  // Collects stats using mi_visit.. also logs, to see logs run the test with:
+  // --vmodule=page_usage_stats_test=1 --logtostderr
+  const auto before = LogMemStats(m_.heap());
+
+  PageUsage p{CollectPageStats::NO, 0.8};
+  for (auto& j : all_objects) {
+    if (j.has_value()) {
+      Defragment(j.value(), &p);
+    }
+  }
+
+  mi_heap_collect(m_.heap(), true);
+  const auto after = LogMemStats(m_.heap());
+
+  EXPECT_LT(after.total_wasted, before.total_wasted);
+  EXPECT_LT(after.total_committed, before.total_committed);
+}
+
+namespace {
+
+void InitBenchMemRes() {
+  static bool initialized = false;
+  if (!initialized) {
+    auto* tlh = mi_heap_get_backing();
+    init_zmalloc_threadlocal(tlh);
+    static MiMemoryResource m{tlh};
+    InitTLStatelessAllocMR(&m);
+    CompactObj::InitThreadLocal(&m);
+    initialized = true;
+  }
+}
+
+}  // namespace
+
+void BM_JSONDefragSerialize(benchmark::State& state) {
```

Comment:
this and the next comparison focused benchmark can be removed in the next pr once the serialization based defrag is cleanup up.